crates/turborepo-lsp/src/lib.rs

```diff
@@ -0,0 +1,858 @@
+    files: Mutex<HashMap<Url, crop::Rope>>,
+    initializer: tokio::sync::watch::Sender<Option<Arc<AsyncMutex<DaemonClient<DaemonConnector>>>>>,
+    daemon: tokio::sync::watch::Receiver<Option<Arc<AsyncMutex<DaemonClient<DaemonConnector>>>>>,
+}
+
+#[tower_lsp::async_trait]
+impl LanguageServer for Backend {
+    async fn initialize(&self, params: InitializeParams) -> LspResult<InitializeResult> {
+        if let Some(uri) = params.root_uri {
+            if uri.scheme() != "file" {
+                return Err(Error::invalid_params("root is not a local file"));
+            }
+
+            let repo_root = uri
+                .to_file_path()
+                .map_err(|_| Error::invalid_params("root is not a local file"))?;
+            let repo_root = repo_root
+                .as_os_str()
+                .to_str()
+                .ok_or(Error::invalid_params("root is not a valid utf-8 path"))?;
+
+            // convert uri file:///absolute-path to AbsoluteSystemPathBuf
+            let repo_root =
+                AbsoluteSystemPathBuf::new(repo_root).expect("file is always an absolute path");
+
+            self.repo_root
+                .lock()
+                .expect("only fails if poisoned")
+                .replace(repo_root.clone());
+
+            let hasher = DaemonRootHasher::new(&repo_root);
+
+            self.client
+                .log_message(
+                    MessageType::INFO,
+                    format!("root uri: {}", hasher.sock_path()),
+                )
+                .await;
+
+            // attempt to connect a few times
+            let daemon = tokio_retry::Retry::spawn(
+                tokio_retry::strategy::FixedInterval::from_millis(100),
+                || {
+                    let connector = DaemonConnector {
+                        can_start_server: true,
+                        can_kill_server: false,
+                        pid_file: hasher.lock_path(),
+                        sock_file: hasher.sock_path(),
+                    };
+                    connector.connect()
+                },
+            )
+            .await;
+
+            let daemon = match daemon {
+                Ok(daemon) => daemon,
+                Err(e) => {
+                    self.client
+                        .log_message(
+                            MessageType::ERROR,
+                            format!("failed to connect to daemon: {}", e),
+                        )
+                        .await;
+                    return Err(Error::internal_error());
+                }
+            };
+
+            self.initializer
+                .send(Some(Arc::new(AsyncMutex::new(daemon))))
+                .expect("there is a receiver");
+        }
+
+        Ok(InitializeResult {
+            server_info: None,
+            capabilities: ServerCapabilities {
+                text_document_sync: Some(TextDocumentSyncCapability::Kind(
+                    TextDocumentSyncKind::INCREMENTAL,
```

Comment:
We only get file diffs sent over so we will need to store our own version of the data (using ropes!)